Windows + HIP build fails (__assert_fail undeclared)

**Describe the bug**

When building a kokkos program on windows with HIP backend, configure succeeds, but build fails with the following errors:
```
C:/Users/WilliamSati/source/repos/kokkos-tutorials/Exercises/dep/Kokkos/core/src\HIP/Kokkos_HIP_Abort.hpp:19:3: error: use of undeclared identifier '__assert_fail'
```
The full build output is here:
[output.txt](https://github.com/user-attachments/files/26906683/output.txt)

This is a similar issue to the one submitted about sycl backend with windows:
https://github.com/kokkos/kokkos/issues/9034

**Please include the following for a minimal reproducer**

1. compilers:
clang 21.0, clang++ 21.0
2. branch hash:
 6a8245644e8dbb03119e20a785d2d230aa39d755
3. operating system version:
Windows 11, x64
4. CMake configure command
Preset:
```
    {
      "name": "base",
      "hidden": true,
      "binaryDir": "${sourceDir}/build/${presetName}"
    },
    {
      "name": "hip",
      "displayName": "HIP (AMD ROCm)",
      "description": "Kokkos HIP backend for AMD GPUs. Uses Ninja. C/C++/HIP must use the same ROCm Clang (preset: clang/clang++); add ROCm bin (e.g. ...\\ROCm\\7.1\\bin) to PATH before cmake. CMake rejects MSVC + HIP on Windows. Default gfx matches Kokkos_ARCH_AMD_GFX1100.",
      "inherits": "base",
      "generator": "Ninja",
      "cacheVariables": {
        "CMAKE_C_COMPILER": { "type": "STRING", "value": "clang" },
        "CMAKE_CXX_COMPILER": { "type": "STRING", "value": "clang++" },
        "CMAKE_HIP_ARCHITECTURES": { "type": "STRING", "value": "gfx1100" },
        "Kokkos_ENABLE_HIP": { "type": "BOOL", "value": "ON" },
        "Kokkos_ENABLE_COMPILE_AS_CMAKE_LANGUAGE": { "type": "BOOL", "value": "ON" },
        "Kokkos_ARCH_AMD_GFX1100": { "type": "BOOL", "value": "ON" }
      }
    },
```
5. Output from CMake configure command
```
PS C:\Users\WilliamSati\source\repos\kokkos-tutorials\Exercises\02\Solution> cmake --preset hip
-- The C compiler identification is Clang 21.0.0 with GNU-like command-line
-- The CXX compiler identification is Clang 21.0.0 with GNU-like command-line
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: C:/Program Files/AMD/ROCm/7.1/bin/clang.exe - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: C:/Program Files/AMD/ROCm/7.1/bin/clang++.exe - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Could NOT find Kokkos (missing: Kokkos_DIR)
-- Setting default Kokkos CXX standard to 20
-- The HIP compiler identification is Clang 21.0.0 with GNU-like command-line
-- Detecting HIP compiler ABI info
-- Detecting HIP compiler ABI info - done
-- Check for working HIP compiler: C:/Program Files/AMD/ROCm/7.1/bin/clang++.exe - skipped
-- Detecting HIP compile features
-- Detecting HIP compile features - done
-- Kokkos version: 5.1.99
-- The project name is: Kokkos
-- Configured git information in C:/Users/WilliamSati/source/repos/kokkos-tutorials/Exercises/02/Solution/build/hip/_deps/kokkos-build/generated/Kokkos_Version_Info.cpp
-- Kokkos is configured for CMake languages HIP compilation (using Clang version 21.0.0)
-- SERIAL backend is being turned on to ensure there is at least one Host space. To change this, you must enable another host execution space and configure with -DKokkos_ENABLE_SERIAL=OFF.
CMake Warning (dev) at C:/Program Files/AMD/ROCm/7.1/lib/cmake/hip/hip-config-amd.cmake:98 (message):
   GPU_TARGETS was not set, and system GPU detection was unsuccsessful.

   The amdgpu-arch tool failed:
   Error: 'Failed to get device count'
   Output: ''

   As a result, --offload-arch will not be set for subsequent
   compilations, and the default architecture
   (gfx906 for dynamic build / gfx942 for static build) will be used
   for compiling device code in C++ language mode

Call Stack (most recent call first):
  C:/Program Files/AMD/ROCm/7.1/lib/cmake/hip/hip-config.cmake:151 (include)
  C:/Users/WilliamSati/source/repos/kokkos-tutorials/Exercises/dep/Kokkos/cmake/kokkos_enable_options.cmake:89 (find_package)
  C:/Users/WilliamSati/source/repos/kokkos-tutorials/Exercises/dep/Kokkos/cmake/kokkos_tribits.cmake:173 (include)
  C:/Users/WilliamSati/source/repos/kokkos-tutorials/Exercises/dep/Kokkos/CMakeLists.txt:216 (kokkos_setup_build_environment)
This warning is for project developers.  Use -Wno-dev to suppress it.

-- Performing Test HIP_CLANG_SUPPORTS_PARALLEL_JOBS
-- Performing Test HIP_CLANG_SUPPORTS_PARALLEL_JOBS - Success
-- Using -std=gnu++20 for C++20 extensions as feature
-- Built-in Execution Spaces:
--     Device Parallel: Kokkos::HIP
--     Host Parallel: NoTypeDefined
--       Host Serial: SERIAL
--
-- Architectures:
--  AMD_GFX1100
CMake Warning (dev) at C:/Program Files/AMD/ROCm/7.1/lib/cmake/hip/hip-config-amd.cmake:98 (message):
   GPU_TARGETS was not set, and system GPU detection was unsuccsessful.

   The amdgpu-arch tool failed:
   Error: 'Failed to get device count'
   Output: ''

   As a result, --offload-arch will not be set for subsequent
   compilations, and the default architecture
   (gfx906 for dynamic build / gfx942 for static build) will be used
   for compiling device code in C++ language mode

Call Stack (most recent call first):
  C:/Program Files/AMD/ROCm/7.1/lib/cmake/hip/hip-config.cmake:151 (include)
  C:/Program Files/CMake/share/cmake-4.0/Modules/CMakeFindDependencyMacro.cmake:78 (find_package)
  C:/Program Files/AMD/ROCm/7.1/lib/cmake/rocprim/rocprim-config.cmake:90 (find_dependency)
  C:/Program Files/CMake/share/cmake-4.0/Modules/CMakeFindDependencyMacro.cmake:78 (find_package)
  C:/Program Files/AMD/ROCm/7.1/lib/cmake/rocthrust/rocthrust-config.cmake:90 (find_dependency)
  C:/Users/WilliamSati/source/repos/kokkos-tutorials/Exercises/dep/Kokkos/cmake/Modules/FindTPLROCTHRUST.cmake:1 (find_package)
  C:/Users/WilliamSati/source/repos/kokkos-tutorials/Exercises/dep/Kokkos/cmake/kokkos_functions.cmake:260 (find_package)
  C:/Users/WilliamSati/source/repos/kokkos-tutorials/Exercises/dep/Kokkos/cmake/kokkos_tpls.cmake:71 (kokkos_import_tpl)
  C:/Users/WilliamSati/source/repos/kokkos-tutorials/Exercises/dep/Kokkos/cmake/kokkos_tribits.cmake:177 (include)
  C:/Users/WilliamSati/source/repos/kokkos-tutorials/Exercises/dep/Kokkos/CMakeLists.txt:216 (kokkos_setup_build_environment)
This warning is for project developers.  Use -Wno-dev to suppress it.

-- Using bundled desul_atomics copy (desul/desul@68f8e83926657f2669712a12e97ec71fd59b72a6)
-- Performing Test KOKKOS_LINK_OPTIONS_CHECK
-- Performing Test KOKKOS_LINK_OPTIONS_CHECK - Failed
CMake Warning at C:/Users/WilliamSati/source/repos/kokkos-tutorials/Exercises/dep/Kokkos/cmake/kokkos_functions.cmake:1007 (message):
  The linker for HIP can not consume flag(s) -DKOKKOS_DEPENDENCE
  -fno-gpu-rdc.  Please check the given configuration.
Call Stack (most recent call first):
  C:/Users/WilliamSati/source/repos/kokkos-tutorials/Exercises/dep/Kokkos/cmake/kokkos_tribits.cmake:286 (kokkos_check_flags)
  C:/Users/WilliamSati/source/repos/kokkos-tutorials/Exercises/dep/Kokkos/cmake/kokkos_tribits.cmake:423 (kokkos_set_library_properties)
  C:/Users/WilliamSati/source/repos/kokkos-tutorials/Exercises/dep/Kokkos/core/src/CMakeLists.txt:131 (kokkos_add_library)


-- Performing Test KOKKOS_COMPILE_OPTIONS_CHECK
-- Performing Test KOKKOS_COMPILE_OPTIONS_CHECK - Failed
CMake Warning at C:/Users/WilliamSati/source/repos/kokkos-tutorials/Exercises/dep/Kokkos/cmake/kokkos_functions.cmake:992 (message):
  The compiler for HIP can not consume flag(s) -mcx16 -fno-gpu-rdc -xhip
  --offload-arch=gfx1100 in combination with the CMAKE_HIP_FLAGS=.  Please
  check the given configuration.
Call Stack (most recent call first):
  C:/Users/WilliamSati/source/repos/kokkos-tutorials/Exercises/dep/Kokkos/cmake/kokkos_tribits.cmake:346 (kokkos_check_flags)
  C:/Users/WilliamSati/source/repos/kokkos-tutorials/Exercises/dep/Kokkos/cmake/kokkos_tribits.cmake:423 (kokkos_set_library_properties)
  C:/Users/WilliamSati/source/repos/kokkos-tutorials/Exercises/dep/Kokkos/core/src/CMakeLists.txt:131 (kokkos_add_library)


-- Using bundled mdspan copy (kokkos/mdspan@5d4eb209c77f4744980c0b0c2af44636cc81b08b)
-- Kokkos Backends: SERIAL;HIP
-- Configuring done (33.8s)
-- Generating done (0.2s)
-- Build files have been written to: C:/Users/WilliamSati/source/repos/kokkos-tutorials/Exercises/02/Solution/build/hip
```
6. Minimum, complete code needed to reproduce the bug
```
#include <Kokkos_Core.hpp>

int main(int argc, char* argv[]) {
  Kokkos::initialize(argc, argv);
  Kokkos::finalize();
  return 0;
}
```
7. Command line needed to reproduce the bug
cmake --build --preset hip
9. `KokkosCore_config.h` header file (generated during the build)

[kokkosCore_config.txt](https://github.com/user-attachments/files/26906681/kokkosCore_config.txt)

10. Please provide any additional relevant error logs

[output.txt](https://github.com/user-attachments/files/26906683/output.txt)

--- a/core/src/HIP/Kokkos_HIP_Abort.hpp
+++ b/core/src/HIP/Kokkos_HIP_Abort.hpp
@@ -5,6 +5,7 @@
 #define KOKKOS_HIP_ABORT_HPP
 
 #include <Kokkos_Macros.hpp>
+#include <Kokkos_Printf.hpp>
 
 #include <hip/hip_runtime.h>
 
@@ -16,7 +17,17 @@ namespace Impl {
 [[noreturn]] __device__ __attribute__((noinline)) inline void hip_abort(
     char const *msg) {
   const char empty[] = "";
+#if defined(_WIN32)
+  // ROCm's Windows device runtime ships no glibc __assert_fail. Emit the
+  // message via device printf (as the SYCL backend does on _MSC_VER), then
+  // abort(). The while(true) guard below preserves [[noreturn]] since abort()
+  // is not itself marked [[noreturn]].
+  (void)empty;
+  Kokkos::printf("Aborting with message %s.\n", msg);
+  abort();
+#else
   __assert_fail(msg, empty, 0, empty);
+#endif
   // This loop is never executed. It's intended to suppress warnings that the
   // function returns, even though it does not. This is necessary because
   // abort() is not marked as [[noreturn]], even though it does not return.

